### PR TITLE
Adding Energy Storage Cache

### PR DESCRIPTION
**Pull request by sourabhdalvi**
Changes :
- [x] New cache type
- [x] Implement correct initial condition values for Hydro Energy Reservoir and Battery 
- [x] Tests for `EnergyStored` Cache
- [x] Test for Cache modified by multiple stages 

This is Copy of PR #370

### CODE REVIEW

File: src/PowerSimulations.jl [hunk 1]
```diff
@@ -80,6 +80,7 @@ export DeviceEnergy
 
 # cache_models
 export TimeStatusChange
+export EnergyStored
```
Comment (jd-lara) [inline src/PowerSimulations.jl]: switch to StoredEnergy

File: src/core/cache.jl [hunk 2]
```diff
@@ -21,4 +21,15 @@ function TimeStatusChange(::Type{T}, name::AbstractString) where {T <: PSY.Devic
     return TimeStatusChange(T, value_array, UpdateRef{JuMP.VariableRef}(T, name))
 end
 
+mutable struct EnergyStored <: AbstractCache
+    device_type::Type{<:PSY.Device}
+    value::JuMP.Containers.DenseAxisArray{Float64}
+    ref::UpdateRef
+end
+
+function EnergyStored(::Type{T}, name::AbstractString) where {T <: PSY.Device}
```
Comment (jd-lara) [inline src/core/cache.jl]: Same here please

File: src/core/initial_conditions.jl [hunk 3]
```diff
@@ -99,6 +99,21 @@ function calculate_ic_quantity(
     return var_value
 end
 
+function calculate_ic_quantity(
+    initial_condition_key::ICKey{DeviceEnergy, T},
```
Comment (jd-lara) [inline src/core/initial_conditions.jl]: EnergyLevel  or energy quantity makes more sense

File: src/core/simulation.jl [hunk 4]
```diff
@@ -425,10 +425,15 @@ function initial_condition_update!(
     ini_cond_vector = get_initial_conditions(stage.internal.psi_container)[ini_cond_key]
     for ic in get_initial_conditions(stage.internal.psi_container)[ini_cond_key]
         name = device_name(ic)
-        interval_chronology = get_stage_interval_chronology(sim, stage.name)
+        interval_chronology =
+            get_stage_interval_chronology(sim.sequence, get_stage_name(sim, stage))
         var_value =
             get_stage_variable(interval_chronology, (stage => stage), name, ic.update_ref)
-        cache = get_cache(stage, ic.cache_key)
+        if isnothing(ic.cache_type)
```
Comment (jd-lara) [inline src/core/simulation.jl]: why this here? the get cache should return a nothing if you run the get_cache function.

Comment (sourabhdalvi) [reply src/core/simulation.jl]: sorry, about that

Comment (sourabhdalvi) [reply src/core/simulation.jl]: Well this function only takes `<: AbstractCache ` but `ic.cache_type` can be a nothing which breaks the that call and this is how you have it for `InterStageChronology`

File: src/PowerSimulations.jl (hunk 1 above)
Comment (sourabhdalvi) [reply src/PowerSimulations.jl]: will do

File: src/core/initial_conditions.jl (hunk 3 above)
Comment (sourabhdalvi) [reply src/core/initial_conditions.jl]: you want me to change DeviceEnergy to EnergyLevel for  InitialConditionType ?

Comment (sourabhdalvi) [reply src/core/initial_conditions.jl]: done

File: src/core/simulation.jl (hunk 4 above)
Comment (jd-lara) [reply src/core/simulation.jl]: ok, just leave it like this. I need to do some refactoring of these updates to make the code cleaner.

**Pull request merged**